### DISCUSSION TITLE
Cloudflare workers

### DISCUSSION THREAD

**Discussion by dantodor** [duckdb/duckdb-wasm#430] [Q&A] [answered]
Has anyone tried to use this within a CF worker ? I think it's a great use case, put your data in some parquet files somewhere ( web server, maybe later even integrate with CF R2 ), run your query engine at the edge.

**Comment by ankoh** [comment]
We'd love to! But the bundle size of Cloudflare workers is unfortunately capped at 1 MB right now, which gets problematic for Wasm very quickly.
Our gzipped wasm module is around 1.8 MB so we cannot bundle the module alongside the worker script today.
(We could probably just fetch the module but we'd have to find out what that means for the startup latency of the worker)
We'd like to explore this a bit more in the following months since CF unbounded workers could be a perfect fit for certain Serverless SQL workloads.

**Reply by dantodor** [reply]
@ankoh It took a while, but I got the limit increased to 2MB on CF for my workers :) .... Would love to give it a try... Any hints where to start ?

**Reply by ankoh** [reply]
Exciting! Can you please join our discord server [here](https://discord.gg/tcvwpjfnZx) and write me a DM? I'd love to chat about this.

**Reply by ilyabo** [reply]
@ankoh You mentioned that there's a gzipped wasm module which is around 1.8 MB. I am using the one from jsdelivr and chrome dev tools say that it's 2.8 MB gzipped. Is it a different one?

![image](https://user-images.githubusercontent.com/351828/149790070-554c3018-fdfa-4b96-8328-7a8f9d419419.png)

**Reply by ankoh** [reply]
@ilyabo brotli compression should bring you down to 1.8. maybe 2.0 in more recent builds.

**Reply by ankoh** [reply]
Right, the jsdelivr ones seem to be served with gzip and are a bit larger.

**Reply by ilyabo** [reply]
Okay I see, thanks André

**Comment by ankoh** [comment, marked as answer]
Update here:
The biggest limitation with cloudflare workers is the fact that cloudflare only implements the Service Worker API.
This hinders us to, for example, use synchronous XHRs that are a strong prerequisite for our web filesystem at the moment.
That might change in the future, but remains a blocker for the time being.

**Reply by danyx23** [reply]
@ankoh can you elaborate a bit on why you need synchronous XHRs? With the availability of the [R2](https://www.cloudflare.com/products/r2/) object storage that is access via cloudflare workers it starts to become extremely appealing to use duckdb to extract/transform data living in a lake in R2 very close to the client. What would be needed to make this work with async Service worker apis?

**Reply by ankoh** [reply]
I fully agree that it would be fantastic but we cannot use asynchronous APIs in the context of blocking I/O calls.
Today, the filesystem abstraction of DuckDB is built against blocking Posix reads and writes which are inherently incompatible to Javascript Promises.
We simply cannot await anything when invoking table functions today.

I only see 3 options that would unblock us with cloudflare:
A) DuckDB gets async I/O, this is a lot of effort on our end that propagates through the entire codebase
B) Cloudflare supports synchronous XHRs
C) Browsers move forward with the wasm stack-switching proposal (https://github.com/WebAssembly/stack-switching)

**Reply by ankoh** [reply]
I had a brief chat with @evanderkoogh on this issue. They solve this with a custom fork of asyncify (sophisticated compiler rewrite pass of emscripten) but I have the very strong feeling that even a patched asyncify fork wont help us here as the abstractions in DuckDB are not as easily rewritable compared to the VDBE of sqlite.
But I have yet to try their fork out to prove that. (unpatched asyncify never terminates)

**Reply by danyx23** [reply]
Ah that makes sense - thanks for the explanation! I can see how that is a tricky issue to resolve. It probably doesn't matter but if a use of asyncify would result in a version of duckdb-wasm that can only read and not write files that would already be a win for such scenarios.

Anyhow, thanks a lot and I appreciate your effort very much!

**Reply by ignoramous** [reply]
@ankoh the size limit for Workers now [stands at 5MB](https://archive.is/lvSAv) (enough to bundle in duckdb-wasm which is around ~3.5MB `gzip`d). Going from the conversation here, synch xhr is the only hold up now? Generally speaking, ignoring the effort involved, would async IO speed up DuckDB? If so, there's at least hope that DuckDB will eventually support it.

**Reply by elithrar** [reply]
(I work at Cloudflare on Workers + Databases)

- Bundle size is an easy one. Increases to that very soon.
- Synchronous XHR much harder, but…
- Wasm stack switching (https://v8.dev/blog/jspi) is something we’re evaluating.

The good thing is that DuckDB works with R2 now. The less good thing is that we’d obviously love to allow folks to run DuckDB on Workers — but once we get unblocked I’ll reach out to Hannes & let y’all know.

**Reply by carlopi** [reply]
Stack switching might be a way around sync XHR, or another potential path is work by @samansmink on resumable pipelines, a big step towards async I/O.

Looking forward for any news you might have, and might come back later with some questions.

**Reply by clementgalopin** [reply]
hi @elithrar, any news regarding the support of duckdb?

**Reply by ShravanSunder** [reply]
As of 2024-04
> A Worker can be up to 10 MB in size after compression on the Workers Paid plan, and up to 1 MB on the Workers Free plan.
The bundle size limits have been increased.   
Snchronous xhr is still not supported, 
 i think workers have first class support for wasm.

**Reply by abhishiv** [reply]
Since duckdb requires sync xhr, i'm wondering if it would be possible to mock that in a durable object? Does anyone know if DO support SharedArrayBuffer & Atomics?

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Atomics/wait

**Reply by tobilg** [reply]
@elithrar would DurableObjects be an alternative that makes more sense? IMO, it'd be a dream to be able to use DuckDB for analytical workload on the edge, as it's possible with SQLite...

**Reply by kimmolinna** [reply]
@kentonv what do you think? Is it possible to support @duckdb with @cloudflare's Durable object in a similar way as #sqlite?

**Reply by kentonv** [reply]
We (Cloudflare) embedded sqlite natively into our runtime, not using Wasm. We are extremely conservative about what kinds of dependencies we're willing to do this with -- sqlite meets a very high bar through the fact that it has been stable, secure, and used by basically everyone for decades. In comparison, duckdb is relatively young and unproven, so embedding it would be a pretty risky commitment for us.

So, I think Wasm is going to have to be the answer here.

It sounds like the sticking point right now is that DuckDB wants to perform synchronous HTTP requests, which Cloudflare Workers does not support (and which has long been deprecated in browsers).

Probably the solution to this problem is JSPI, aka Wasm stack switching. This allows Wasm applications to make synchronous-looking calls that are actually implemented by asynchronous JavaScript, by temporarily suspending the Wasm call stack. JSPI has been in development for a very long time, but [it looks like](https://issues.chromium.org/issues/42202153) it's finally rolling out by default in Chrome, which means we'll enable it in Workers soon (like, in the next couple weeks -- funny coincidence).

(Historically, you could also use [Asyncify](https://emscripten.org/docs/porting/asyncify.html) to convert Wasm to be asynchronous without explicit runtime support, although it comes with a performance hit. JSPI should be much better.)

**Reply by carlopi** [reply]
@kentonv: Me being lazy and taking advantage of you being in the conversation, but it would be possibly to mock synchronous HTTP requests via threads locking + perform async request in a separate thread? Are Wasm threads supported in Cloudflare workers?

**Reply by kentonv** [reply]
No, we do not support threads in Cloudflare Workers.

(They wouldn't be a good fit for our platform since we really want heavy compute tasks to be distributed across multiple _machines_. Having one Worker take multiple cores on a single machine would actually be problematic as it would make load balancing much harder. Also there are Spectre concerns.)

**Reply by ignoramous** [reply]
Me being lazy too, but can Workers support full-duplex requests, too? https://github.com/posener/h2conn Comes in handy to pool connections from a client if Workers was backed by a db.

**Reply by kentonv** [reply]
@ignoramous haha this seems a little off topic, maybe file a feature request on workerd? :) (Actually, though, workerd itself supports full duplex even on HTTP/1.1, but the rest of the Cloudflare HTTP proxy stack does not...)

**Reply by kimmolinna** [reply]
@kentonv thank you very much for your answer. I fully understand your policy for the  production environment but how about a beta platform? 😉 I know that there are many capable duckdb persons (like @tobilg) who could help you with a beta implementation.

**Reply by jasnell** [reply]
Workers really has no separate "beta platform" and if it did it would be subject to the same policy. We can explore options for implementing some form of support but the kind of multiple thread and blocking http requests discussed so far is not currently possible. Note that we are working on implementing node:fs APIs in workers using a virtual in memory file system. Would a variation on the node binding here be a possibility?

**Reply by tobilg** [reply]
<img width="1012" height="601" alt="duckdb-on-workers" src="https://github.com/user-attachments/assets/ae4f606b-f50d-458a-bf21-64d38bd4a6da" />

I have a single-threaded version of DuckDB running on Workers (Pro plan due to size limits). It includes statically linked Parquet/JSON/httpfs extensions, meaning that it can query data on R2. I'll release a npm package for this soon. 

cc @kentonv / @elithrar

**Reply by tobilg** [reply]
FYI: https://www.npmjs.com/package/@ducklings/workers

**Comment by vcknorket** [comment]
Curious is this still in the works? Any plans to enable this?

**Reply by ockhamlabs** [reply]
@elithrar any update on this? @ankoh are u able to let us know how you were able to get this working?

**Comment by mustafahasankhan** [comment]
Folks @ankoh / @elithrar 
1. Is there now an async-only version of DuckDB-WASM that doesn't rely on synchronous XHR?
2. Has anyone successfully used DuckDB with Durable Objects, which might provide different capabilities?
3. Are there any workarounds using Web Workers or other techniques to run DuckDB in this environment?

**Reply by ignoramous** [reply]
> ... other techniques to run DuckDB in this environment?

May be possible with containers (releasing June): https://blog.cloudflare.com/cloudflare-containers-coming-2025/ ([mirror](https://archive.vn/Sto41)).

**Reply by tobilg** [reply]
My 2ct:

1. Afaik it's only possible the use in the browser with COI enabled (https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Headers/Permissions-Policy/cross-origin-isolated)
2. I doubt that this would be possible, as only SQLite is supported currently
3. I doubt that because of the plain Workers size limits of 3MB (free) and 10MB (paid)

I'm on the waitlist for Containers preview access, if I get it I might report back.

**Reply by archiewood** [reply]
Containers is out!

**Reply by vcknorket** [reply]
Link?

**Reply by archiewood** [reply]
https://developers.cloudflare.com/containers/

**Reply by vcknorket** [reply]
this works for duckdb?

**Reply by tobilg** [reply]
> this works for duckdb?

See https://github.com/tobilg/cloudflare-duckdb

**Reply by azu** [reply]
> Is there now an async-only version of DuckDB-WASM that doesn't rely on synchronous XHR?

https://tobilg.com/posts/custom-duckdb-wasm-builds-for-cloudflare-workers/
https://github.com/tobilg/ducklings

Ducklings seems to solve this problem.

**Comment by earonesty** [comment]
use this stack:

  icebird
  hyparquet
  hyparquet-writer
  
  problem solved
